controllers/namespace/namespace_controller.go

```diff
@@ -0,0 +1,219 @@
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// annotateOpenshiftNamespaces annotates all the namespaces starting with openshift-*
+func (r *NamespaceReconciler) annotateOpenshiftNamespaces() (bool, error) {
+	nsList := &corev1.NamespaceList{}
+	err := r.Client.List(r.ctx, nsList)
+	if err != nil {
+		r.Log.Error(err, "Failed to list Namespaces")
+		return false, err
+	}
+
+	openshiftNamespaces := []corev1.Namespace{}
+	for _, ns := range nsList.Items {
+		if strings.HasPrefix(ns.Name, "openshift-") {
+			openshiftNamespaces = append(openshiftNamespaces, ns)
+		}
+	}
+
+	return r.annotateNamespaces(openshiftNamespaces)
+}
+
+// annotateNamespaces annotate the provided namespace starting but, ignores if
+// it has annotation "'skipReclaimspaceSchedule':'true'" set
+func (r *NamespaceReconciler) annotateNamespaces(opNs []corev1.Namespace) (bool, error) {
+	updated := false
+	for _, ns := range opNs {
+		if shouldSkipAnnotation(ns) {
+			continue
+		}
+		if ns.Annotations == nil {
+			ns.Annotations = make(map[string]string)
+		}
+		ns.Annotations["reclaimspace.csiaddons.openshift.io/schedule"] = "@weekly"
+
+		err := r.Client.Update(r.ctx, &ns)
+		if err != nil {
+			r.Log.Error(err, fmt.Sprintf("Failed to annotate namespace %s", ns.Name))
+			return false, err
+		}
+		r.Log.Info(fmt.Sprintf("Successfully annotated namespace: %s", ns.Name))
+		updated = true
+	}
+	return updated, nil
+}
+
+// shouldSkipAnnotation checks if we should skip adding new annotations to the namespace
+func shouldSkipAnnotation(ns corev1.Namespace) bool {
+	if !ns.GetDeletionTimestamp().IsZero() || ns.Annotations["skipReclaimspaceSchedule"] == "true" ||
+		ns.Annotations["reclaimspace.csiaddons.openshift.io/schedule"] != "" {
+		return true
+	}
+	return false
+}
+
+// restartCSIAddonsPod restarts the CSIAddons pod by deleting all of its instances.
+func (r *NamespaceReconciler) restartCSIAddonsPod() error {
+	podList := &corev1.PodList{}
+	listOpts := []client.ListOption{
+		client.InNamespace(os.Getenv(util.OperatorNamespaceEnvVar)),
+		client.MatchingLabels{"app.kubernetes.io/name": "csi-addons"},
+	}
+	err := r.Client.List(r.ctx, podList, listOpts...)
+	if err != nil {
+		r.Log.Error(err, "Failed to get list of pods with label 'csi-addons'")
+		return err
+	}
+
+	// If no csi-addons pod is present
+	if len(podList.Items) == 0 {
+		r.Log.Error(err, "No csi-addons pod found")
+		return nil
+	}
+
+	for _, csiAddonsPod := range podList.Items {
+		err = r.Client.Delete(r.ctx, &csiAddonsPod, client.GracePeriodSeconds(0))
+		if err != nil {
+			r.Log.Error(err, "Failed to delete csi-addons pod")
+			return err
+		}
+	}
+
+	return nil
+}
+
+// SetupWithManager sets up a controller with a manager
+func (r *NamespaceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	generationChangedPredicate := predicate.GenerationChangedPredicate{}
+
+	predicateFunc := func(obj runtime.Object) bool {
```

Comment:
Can we add more predicates to skip adding annotations instead of doing it in reconcile? If namespace has the skip annotation or annotation already exists, we don't need to trigger reconcile.

Reply:
Sure, though I have kept the `skipAnnotation` check inside the reconcile as well, for the `once.Do` func call.